### PR TITLE
[MP Rewrite Part 3] Add PlanRequest and the planner node.

### PR DESCRIPTION
**Pull request by kylestach**
This also overhauls the MotionCommand structure.

### CODE REVIEW

File: soccer/Processor.cpp [hunk 1]
```diff
@@ -252,6 +252,11 @@ void Processor::run() {
         // TODO(Kyle): This is dead code for now. Once everything is ported over
         // to modules we can delete the if (false), but for now we still have to
         // update things manually.
+#if 0
+        for (auto* node : _nodes) {
+            node->run();
+        }
+#endif
```
Comment (oswinso) [inline soccer/Processor.cpp]: I think you should just remove this directly

File: soccer/planning/planner/MotionCommand.hpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * An empty "do-nothing" motion command.
+ */
+struct EmptyCommand {};
+
+/**
+ * Move to a particular target with a particular velocity, avoiding obstacles.
+ */
+// note: as of now the heading and angular velocity are ignored
+// TODO(Kyle) use heading and angular velocity or change this to a
+// LinearMotionInstant
```
Comment (oswinso) [inline soccer/planning/planner/MotionCommand.hpp]: Issue num

File: soccer/planning/planner/MotionCommand.hpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+#include <Geometry2d/Pose.hpp>
+#include <WorldState.hpp>
+#include <variant>
+
+#include "Utils.hpp"
+#include "planning/Instant.hpp"
+#include "planning/Trajectory.hpp"
+
+namespace Planning {
+// todo(Ethan) discuss whether we want to have settle with bounce
```
Comment (oswinso) [inline soccer/planning/planner/MotionCommand.hpp:13]: Issue num

**Review by oswinso — CHANGES_REQUESTED** [review]
More issue nums

File: soccer/planning/planner/MotionCommand.hpp (hunk 3 above)
Comment (kylestach) [reply soccer/planning/planner/MotionCommand.hpp:13]: This is resolved, I'll delete it.

**Review by ghost — APPROVED** [review]

File: soccer/planning/planner/MotionCommand.hpp (hunk 2 above)
Comment (kylestach) [reply soccer/planning/planner/MotionCommand.hpp]: #1505

File: soccer/Robot.cpp [hunk 4]
```diff
@@ -178,8 +170,11 @@ void OurRobot::moveTuning(Geometry2d::Point goal, float endSpeed) {
              << ", " << goal.y() << ")" << endl;
     }
 
-    intent().motion_command = std::make_unique<Planning::TuningPathCommand>(
-        MotionInstant(goal, (goal - pos()).normalized() * endSpeed));
+    Geometry2d::Point targetPoint = goal;
+    Geometry2d::Point targetVel = (goal - pos()).normalized() * endSpeed;
+    RobotInstant goal_instant{Geometry2d::Pose{targetPoint, 0},
+                              Geometry2d::Twist{targetVel, 0}, RJ::Time{0s}};
+    setMotionCommand(Planning::PathTargetCommand{goal_instant});
```
Comment (JNeiger) [inline soccer/Robot.cpp:177]: This is going from TuningPathCommand -> PathTargetCommand. I haven't looked, but I imagine they're different

File: soccer/Robot.cpp [hunk 5]
```diff
@@ -196,28 +191,30 @@ void OurRobot::move(Geometry2d::Point goal, Geometry2d::Point endVelocity) {
              << goal.y() << ")" << std::endl;
     }
 
-    intent().motion_command = std::make_unique<Planning::PathTargetCommand>(
-        MotionInstant(goal, endVelocity));
+    RobotInstant goal_instant;
+    goal_instant.pose = Pose{goal, 0};
```
Comment (JNeiger) [inline soccer/Robot.cpp]: This sets the heading target to 0 for every normal move?

File: soccer/Robot.cpp [hunk 6]
```diff
@@ -196,28 +191,30 @@ void OurRobot::move(Geometry2d::Point goal, Geometry2d::Point endVelocity) {
...
+    goal_instant.velocity = Twist{endVelocity, 0};
+    setMotionCommand(Planning::PathTargetCommand{goal_instant});
 
     _cmdText << "move(" << goal.x() << ", " << goal.y() << ")" << endl;
     _cmdText << "endVelocity(" << endVelocity.x() << ", " << endVelocity.y()
              << ")" << endl;
 }
 
-void OurRobot::settle(const std::optional<Point>& target) {
+void OurRobot::settle() {
```
Comment (JNeiger) [inline soccer/Robot.cpp]: Settle one touch thing

File: soccer/planning/PlannerNode.cpp [hunk 7]
```diff
@@ -1,99 +1,132 @@
...
+            planned.at(shell) = &trajectories->at(shell);
+            continue;
+        }
 
-    // Visualize obstacles
-    for (const auto& shape : global_obstacles.global_obstacles.shapes()) {
-        context_->debug_drawer.drawShape(shape, Qt::black, "Global Obstacles");
+        RobotInstant start{robot.pose, robot.velocity, robot.timestamp};
+
+        // TODO(Kyle): Put motion constraints in intent.
```
Comment (JNeiger) [inline soccer/planning/PlannerNode.cpp]: todo issue #

File: soccer/planning/PlannerNode.hpp [hunk 8]
```diff
@@ -1,40 +1,36 @@
...
-    Geometry2d::ShapeSet global_obstacles_with_goal_zones;
+class PlannerForRobot {
+public:
+    PlannerForRobot();
+
+    Trajectory PlanForRobot(const Planning::PlanRequest& request);
+
+private:
+    std::vector<std::unique_ptr<Planner>> planners_;
+    int planner_idx_;
```
Comment (JNeiger) [inline soccer/planning/PlannerNode.hpp]: What is this used for?

File: soccer/planning/PlannerNode.cpp [hunk 9]
```diff
@@ -1,99 +1,132 @@
...
-    const Geometry2d::ShapeSet goal_zone_obstacles =
-        context_->goalZoneObstacles;
-    Geometry2d::ShapeSet global_obstacles_with_goal_zones = global_obstacles;
-    global_obstacles_with_goal_zones.add(goal_zone_obstacles);
+PlannerForRobot::PlannerForRobot() : planner_idx_(-1) {
+    planners_.push_back(std::make_unique<PathTargetPlanner>());
+    planners_.push_back(std::make_unique<SettlePlanner>());
+    planners_.push_back(std::make_unique<CollectPlanner>());
+    planners_.push_back(std::make_unique<LineKickPlanner>());
+    planners_.push_back(std::make_unique<PivotPathPlanner>());
```
Comment (JNeiger) [inline soccer/planning/PlannerNode.cpp:82]: Does this order matter in this section? Or is it just the empty planner must be last?

File: soccer/Robot.cpp (hunk 5 above)
Comment (kylestach) [reply soccer/Robot.cpp]: Heading for this was ignored. I'm changing it to a linear motion instant to represent that.

File: soccer/Robot.cpp (hunk 4 above)
Comment (kylestach) [reply soccer/Robot.cpp:177]: We're removing the old `TuningPathCommand` for now. I'll add a TODO for it.

File: soccer/planning/PlannerNode.hpp (hunk 8 above)
Comment (kylestach) [reply soccer/planning/PlannerNode.hpp]: Left in accidentally. Just deleted it.

File: soccer/planning/PlannerNode.cpp (hunk 9 above)
Comment (kylestach) [reply soccer/planning/PlannerNode.cpp:82]: Yeah. It only matters when multiple planners could be used to plan for the same command type. Right now, the only place that happens is with the empty/escape obstacles planner.

**Review by JNeiger — APPROVED** [review]
LGTM

**Review by oswinso — APPROVED** [review]
lgtm

**Pull request merged**